diff -ur git.org/data/ModemManager.service.in git/data/ModemManager.service.in
--- git.org/data/ModemManager.service.in	2024-08-21 10:19:23.397003375 +0000
+++ git/data/ModemManager.service.in	2024-08-21 10:30:33.143987891 +0000
@@ -7,7 +7,7 @@
 [Service]
 Type=dbus
 BusName=org.freedesktop.ModemManager1
-ExecStart=@sbindir@/ModemManager
+ExecStart=@sbindir@/ModemManager --debug
 StandardError=null
 Restart=on-abort
 CapabilityBoundingSet=CAP_SYS_ADMIN CAP_NET_ADMIN
