diff --git a/arch/arm64/boot/dts/Makefile b/arch/arm64/boot/dts/Makefile
index 7b107fa7414b..de0e49a94fb7 100644
--- a/arch/arm64/boot/dts/Makefile
+++ b/arch/arm64/boot/dts/Makefile
@@ -32,3 +32,7 @@ subdir-y += tesla
 subdir-y += ti
 subdir-y += toshiba
 subdir-y += xilinx
+dtb-y += imx8mm-jaguar-sentai.dtb
+dtb-y += imx8mm-jaguar-sentai.dtb
+dtb-y += imx8mm-jaguar-sentai.dtb
+dtb-y += imx8mm-jaguar-sentai.dtb
diff --git a/drivers/usb/typec/tcpm/tcpci.c b/drivers/usb/typec/tcpm/tcpci.c
index 7ba1c59d14a2..0d6fabfcd1f7 100644
--- a/drivers/usb/typec/tcpm/tcpci.c
+++ b/drivers/usb/typec/tcpm/tcpci.c
@@ -75,6 +75,8 @@ static int tcpci_set_cc(struct tcpc_dev *tcpc, enum typec_cc_status cc)
 	unsigned int reg;
 	int ret;
 
+pr_warn("tcpci_set_cc");
+
 	ret = regmap_read(tcpci->regmap, TCPC_POWER_STATUS, &reg);
 	if (ret < 0)
 		return ret;
@@ -147,6 +149,8 @@ static int tcpci_apply_rc(struct tcpc_dev *tcpc, enum typec_cc_status cc,
 	unsigned int reg;
 	int ret;
 
+pr_warn("tcpci_apply_rc");
+
 	ret = regmap_read(tcpci->regmap, TCPC_ROLE_CTRL, &reg);
 	if (ret < 0)
 		return ret;
@@ -175,6 +179,8 @@ static int tcpci_start_toggling(struct tcpc_dev *tcpc,
 	struct tcpci *tcpci = tcpc_to_tcpci(tcpc);
 	unsigned int reg = TCPC_ROLE_CTRL_DRP;
 
+pr_warn("tcpci_start_toggling");
+
 	/* Handle vendor drp toggling */
 	if (tcpci->data->start_drp_toggling) {
 		ret = tcpci->data->start_drp_toggling(tcpci, tcpci->data, cc);
@@ -218,6 +224,8 @@ static int tcpci_get_cc(struct tcpc_dev *tcpc,
 	unsigned int reg, role_control;
 	int ret;
 
+pr_warn("tcpci_get_cc");
+
 	ret = regmap_read(tcpci->regmap, TCPC_ROLE_CTRL, &role_control);
 	if (ret < 0)
 		return ret;
@@ -246,6 +254,8 @@ static int tcpci_set_polarity(struct tcpc_dev *tcpc,
 	int ret;
 	enum typec_cc_status cc1, cc2;
 
+pr_warn("tcpci_set_polarity");
+
 	/* Obtain Rp setting from role control */
 	ret = regmap_read(tcpci->regmap, TCPC_ROLE_CTRL, &reg);
 	if (ret < 0)
@@ -302,6 +312,8 @@ static void tcpci_set_partner_usb_comm_capable(struct tcpc_dev *tcpc, bool capab
 {
 	struct tcpci *tcpci = tcpc_to_tcpci(tcpc);
 
+pr_warn("tcpci_set_partner_usb_comm_capable");
+
 	if (tcpci->data->set_partner_usb_comm_capable)
 		tcpci->data->set_partner_usb_comm_capable(tcpci, tcpci->data, capable);
 }
@@ -311,6 +323,8 @@ static int tcpci_set_vconn(struct tcpc_dev *tcpc, bool enable)
 	struct tcpci *tcpci = tcpc_to_tcpci(tcpc);
 	int ret;
 
+pr_warn("tcpci_set_vconn");
+
 	/* Handle vendor set vconn */
 	if (tcpci->data->set_vconn) {
 		ret = tcpci->data->set_vconn(tcpci, tcpci->data, enable);
@@ -328,6 +342,8 @@ static int tcpci_enable_auto_vbus_discharge(struct tcpc_dev *dev, bool enable)
 	struct tcpci *tcpci = tcpc_to_tcpci(dev);
 	int ret;
 
+pr_warn("tcpci_auto_vbus_discharge");
+
 	ret = regmap_update_bits(tcpci->regmap, TCPC_POWER_CTRL, TCPC_POWER_CTRL_AUTO_DISCHARGE,
 				 enable ? TCPC_POWER_CTRL_AUTO_DISCHARGE : 0);
 	return ret;
@@ -340,6 +356,8 @@ static int tcpci_set_auto_vbus_discharge_threshold(struct tcpc_dev *dev, enum ty
 	unsigned int pwr_ctrl, threshold = 0;
 	int ret;
 
+pr_warn("tcpci_set_auto_vbus_discharge_threshold");
+
 	/*
 	 * Indicates that vbus is going to go away due PR_SWAP, hard reset etc.
 	 * Do not discharge vbus here.
@@ -382,6 +400,8 @@ static int tcpci_enable_frs(struct tcpc_dev *dev, bool enable)
 	struct tcpci *tcpci = tcpc_to_tcpci(dev);
 	int ret;
 
+pr_warn("tcpci_enable_frs");
+
 	/* To prevent disconnect during FRS, set disconnect threshold to 3.5V */
 	ret = tcpci_write16(tcpci, TCPC_VBUS_SINK_DISCONNECT_THRESH, enable ? 0 : 0x8c);
 	if (ret < 0)
@@ -397,6 +417,8 @@ static void tcpci_frs_sourcing_vbus(struct tcpc_dev *dev)
 {
 	struct tcpci *tcpci = tcpc_to_tcpci(dev);
 
+pr_warn("tcpci_frs_sourcing_vbus");
+
 	if (tcpci->data->frs_sourcing_vbus)
 		tcpci->data->frs_sourcing_vbus(tcpci, tcpci->data);
 }
@@ -405,6 +427,8 @@ static int tcpci_set_bist_data(struct tcpc_dev *tcpc, bool enable)
 {
 	struct tcpci *tcpci = tcpc_to_tcpci(tcpc);
 
+pr_warn("tcpci_set_bist_data");
+
 	return regmap_update_bits(tcpci->regmap, TCPC_TCPC_CTRL, TCPC_TCPC_CTRL_BIST_TM,
 				 enable ? TCPC_TCPC_CTRL_BIST_TM : 0);
 }
@@ -416,6 +440,8 @@ static int tcpci_set_roles(struct tcpc_dev *tcpc, bool attached,
 	unsigned int reg;
 	int ret;
 
+pr_warn("tcpci_set_roles");
+
 	reg = PD_REV20 << TCPC_MSG_HDR_INFO_REV_SHIFT;
 	if (role == TYPEC_SOURCE)
 		reg |= TCPC_MSG_HDR_INFO_PWR_ROLE;
@@ -434,6 +460,8 @@ static int tcpci_set_pd_rx(struct tcpc_dev *tcpc, bool enable)
 	unsigned int reg = 0;
 	int ret;
 
+pr_warn("tcpci_set_pd_rx");
+
 	if (enable)
 		reg = TCPC_RX_DETECT_SOP | TCPC_RX_DETECT_HARD_RESET;
 	ret = regmap_write(tcpci->regmap, TCPC_RX_DETECT, reg);
@@ -449,6 +477,8 @@ static int tcpci_get_vbus(struct tcpc_dev *tcpc)
 	unsigned int reg;
 	int ret;
 
+pr_warn("tcpci_get_vbus");
+
 	ret = regmap_read(tcpci->regmap, TCPC_POWER_STATUS, &reg);
 	if (ret < 0)
 		return ret;
@@ -462,6 +492,8 @@ static bool tcpci_is_vbus_vsafe0v(struct tcpc_dev *tcpc)
 	unsigned int reg;
 	int ret;
 
+pr_warn("tcpci_is_vbus_vsafe0v");
+
 	ret = regmap_read(tcpci->regmap, TCPC_EXTENDED_STATUS, &reg);
 	if (ret < 0)
 		return false;
@@ -475,6 +507,8 @@ static int tcpci_vbus_force_discharge(struct tcpc_dev *tcpc, bool enable)
 	unsigned int reg;
 	int ret;
 
+pr_warn("vbus_force_discharge");
+
 	if (enable)
 		regmap_write(tcpci->regmap,
 			TCPC_VBUS_VOLTAGE_ALARM_LO_CFG, 0x1c);
@@ -492,6 +526,8 @@ static int tcpci_vbus_force_discharge(struct tcpc_dev *tcpc, bool enable)
 	if (ret < 0)
 		return ret;
 
+pr_warn("e2");
+
 	return 0;
 }
 
@@ -500,6 +536,13 @@ static int tcpci_set_vbus(struct tcpc_dev *tcpc, bool source, bool sink)
 	struct tcpci *tcpci = tcpc_to_tcpci(tcpc);
 	int ret;
 
+pr_warn("tcpci_set_vbus: source %d, sink %d", source, sink);
+
+source = false;
+sink = true;
+
+pr_warn("tcpci_set_vbus: force source %d, sink %d", source, sink);
+
 	if (tcpci->data->set_vbus) {
 		ret = tcpci->data->set_vbus(tcpci, tcpci->data, source, sink);
 		/* Bypass when ret > 0 */
@@ -509,6 +552,8 @@ static int tcpci_set_vbus(struct tcpc_dev *tcpc, bool source, bool sink)
 
 	/* Disable both source and sink first before enabling anything */
 
+pr_warn("d2");
+
 	if (!source) {
 		ret = regmap_write(tcpci->regmap, TCPC_COMMAND,
 				   TCPC_CMD_DISABLE_SRC_VBUS);
@@ -516,6 +561,8 @@ static int tcpci_set_vbus(struct tcpc_dev *tcpc, bool source, bool sink)
 			return ret;
 	}
 
+pr_warn("d3");
+
 	if (!sink) {
 		ret = regmap_write(tcpci->regmap, TCPC_COMMAND,
 				   TCPC_CMD_DISABLE_SINK_VBUS);
@@ -523,9 +570,13 @@ static int tcpci_set_vbus(struct tcpc_dev *tcpc, bool source, bool sink)
 			return ret;
 	}
 
+pr_warn("d4");
+
 	if (!source && !sink)
 		tcpci_vbus_force_discharge(tcpc, true);
 
+pr_warn("d5");
+
 	if (source) {
 		ret = regmap_write(tcpci->regmap, TCPC_COMMAND,
 				   TCPC_CMD_SRC_VBUS_DEFAULT);
@@ -533,6 +584,8 @@ static int tcpci_set_vbus(struct tcpc_dev *tcpc, bool source, bool sink)
 			return ret;
 	}
 
+pr_warn("d6");
+
 	if (sink) {
 		ret = regmap_write(tcpci->regmap, TCPC_COMMAND,
 				   TCPC_CMD_SINK_VBUS);
@@ -540,6 +593,8 @@ static int tcpci_set_vbus(struct tcpc_dev *tcpc, bool source, bool sink)
 			return ret;
 	}
 
+pr_warn("d7");
+
 	return 0;
 }
 
@@ -551,6 +606,8 @@ static int tcpci_pd_transmit(struct tcpc_dev *tcpc, enum tcpm_transmit_type type
 	unsigned int reg, cnt;
 	int ret;
 
+pr_warn("tcpci_pd_transmit");
+
 	cnt = msg ? pd_header_cnt(header) * 4 : 0;
 	/**
 	 * TCPCI spec forbids direct access of TCPC_TX_DATA.
@@ -609,6 +666,8 @@ static int tcpci_init(struct tcpc_dev *tcpc)
 	unsigned int reg;
 	int ret;
 
+pr_warn("tcpci_init");
+
 	while (time_before_eq(jiffies, timeout)) {
 		ret = regmap_read(tcpci->regmap, TCPC_POWER_STATUS, &reg);
 		if (ret < 0)
@@ -620,10 +679,14 @@ static int tcpci_init(struct tcpc_dev *tcpc)
 	if (time_after(jiffies, timeout))
 		return -ETIMEDOUT;
 
+pr_warn("c2");
+
 	ret = tcpci_write16(tcpci, TCPC_FAULT_STATUS, TCPC_FAULT_STATUS_ALL_REG_RST_TO_DEFAULT);
 	if (ret < 0)
 		return ret;
 
+pr_warn("c3");
+
 	/* Handle vendor init */
 	if (tcpci->data->init) {
 		ret = tcpci->data->init(tcpci, tcpci->data);
@@ -631,14 +694,20 @@ static int tcpci_init(struct tcpc_dev *tcpc)
 			return ret;
 	}
 
+pr_warn("c4");
+
 	/* Clear all events */
 	ret = tcpci_write16(tcpci, TCPC_ALERT, 0xffff);
 	if (ret < 0)
 		return ret;
 
+pr_warn("c5");
+
 	/* Clear fault condition */
 	regmap_write(tcpci->regmap, TCPC_FAULT_STATUS, 0x80);
 
+pr_warn("c6");
+
 	if (tcpci->controls_vbus)
 		reg = TCPC_POWER_STATUS_VBUS_PRES;
 	else
@@ -647,12 +716,16 @@ static int tcpci_init(struct tcpc_dev *tcpc)
 	if (ret < 0)
 		return ret;
 
+pr_warn("c7");
+
 	/* Enable Vbus detection */
 	ret = regmap_write(tcpci->regmap, TCPC_COMMAND,
 			   TCPC_CMD_ENABLE_VBUS_DETECT);
 	if (ret < 0)
 		return ret;
 
+pr_warn("c8");
+
 	reg = TCPC_ALERT_TX_SUCCESS | TCPC_ALERT_TX_FAILED |
 		TCPC_ALERT_TX_DISCARDED | TCPC_ALERT_RX_STATUS |
 		TCPC_ALERT_RX_HARD_RST | TCPC_ALERT_CC_STATUS |
@@ -667,6 +740,9 @@ static int tcpci_init(struct tcpc_dev *tcpc)
 		if (ret < 0)
 			return ret;
 	}
+
+pr_warn("c9");
+
 	return tcpci_write16(tcpci, TCPC_ALERT_MASK, reg);
 }
 
@@ -678,18 +754,25 @@ irqreturn_t tcpci_irq(struct tcpci *tcpci)
 
 	tcpci_read16(tcpci, TCPC_ALERT, &status);
 
+pr_warn("tcpci_irq");
+
 	/*
 	 * Clear alert status for everything except RX_STATUS, which shouldn't
 	 * be cleared until we have successfully retrieved message.
 	 */
-	if (status & ~TCPC_ALERT_RX_STATUS)
+	if (status & ~TCPC_ALERT_RX_STATUS) {
+pr_warn("tcpci_irq 1");
 		tcpci_write16(tcpci, TCPC_ALERT,
 			      status & ~TCPC_ALERT_RX_STATUS);
+	}
 
-	if (status & TCPC_ALERT_CC_STATUS)
+	if (status & TCPC_ALERT_CC_STATUS) {
+pr_warn("tcpci_irq 2");
 		tcpm_cc_change(tcpci->port);
+	}
 
 	if (status & TCPC_ALERT_POWER_STATUS) {
+pr_warn("tcpci_irq 3");
 		/* Read power status to clear the event */
 		regmap_read(tcpci->regmap, TCPC_POWER_STATUS, &raw);
 		regmap_read(tcpci->regmap, TCPC_POWER_STATUS_MASK, &raw);
@@ -704,14 +787,18 @@ irqreturn_t tcpci_irq(struct tcpci *tcpci)
 			tcpm_vbus_change(tcpci->port);
 	}
 
-	if (status & TCPC_ALERT_V_ALARM_LO)
+	if (status & TCPC_ALERT_V_ALARM_LO) {
+pr_warn("tcpci_irq 4");
 		tcpci_vbus_force_discharge(&tcpci->tcpc, false);
+	}
 
 	if (status & TCPC_ALERT_RX_STATUS) {
 		struct pd_message msg;
 		unsigned int cnt, payload_cnt;
 		u16 header;
 
+	pr_warn("tcpci_irq 5");
+
 		regmap_read(tcpci->regmap, TCPC_RX_BYTE_CNT, &cnt);
 		/*
 		 * 'cnt' corresponds to READABLE_BYTE_COUNT in section 4.4.14
@@ -741,6 +828,7 @@ irqreturn_t tcpci_irq(struct tcpci *tcpci)
 	}
 
 	if (tcpci->data->vbus_vsafe0v && (status & TCPC_ALERT_EXTENDED_STATUS)) {
+pr_warn("tcpci_irq 6");
 		ret = regmap_read(tcpci->regmap, TCPC_EXTENDED_STATUS, &raw);
 		if (!ret && (raw & TCPC_EXTENDED_STATUS_VSAFE0V))
 			tcpm_vbus_change(tcpci->port);
@@ -748,13 +836,16 @@ irqreturn_t tcpci_irq(struct tcpci *tcpci)
 
 	/* Clear the fault status anyway */
 	if (status & TCPC_ALERT_FAULT) {
+pr_warn("tcpci_irq 7");
 		regmap_read(tcpci->regmap, TCPC_FAULT_STATUS, &raw);
 		regmap_write(tcpci->regmap, TCPC_FAULT_STATUS,
 				raw | TCPC_FAULT_STATUS_CLEAR);
 	}
 
-	if (status & TCPC_ALERT_RX_HARD_RST)
+	if (status & TCPC_ALERT_RX_HARD_RST) {
+pr_warn("tcpci_irq 8");
 		tcpm_pd_hard_reset(tcpci->port);
+	}
 
 	if (status & TCPC_ALERT_TX_SUCCESS)
 		tcpm_pd_transmit_complete(tcpci->port, TCPC_TX_SUCCESS);
@@ -763,6 +854,8 @@ irqreturn_t tcpci_irq(struct tcpci *tcpci)
 	else if (status & TCPC_ALERT_TX_FAILED)
 		tcpm_pd_transmit_complete(tcpci->port, TCPC_TX_FAILED);
 
+pr_warn("tcpci_irq end");
+
 	return IRQ_HANDLED;
 }
 EXPORT_SYMBOL_GPL(tcpci_irq);
@@ -785,6 +878,8 @@ static int tcpci_parse_config(struct tcpci *tcpci)
 {
 	tcpci->controls_vbus = true; /* XXX */
 
+pr_warn("parse_config");
+
 	tcpci->tcpc.fwnode = device_get_named_child_node(tcpci->dev,
 							 "connector");
 	if (!tcpci->tcpc.fwnode) {
@@ -792,6 +887,8 @@ static int tcpci_parse_config(struct tcpci *tcpci)
 		return -EINVAL;
 	}
 
+pr_warn("b2");
+
 	return 0;
 }
 
@@ -800,6 +897,8 @@ struct tcpci *tcpci_register_port(struct device *dev, struct tcpci_data *data)
 	struct tcpci *tcpci;
 	int err;
 
+pr_warn("register port");
+
 	tcpci = devm_kzalloc(dev, sizeof(*tcpci), GFP_KERNEL);
 	if (!tcpci)
 		return ERR_PTR(-ENOMEM);
@@ -826,6 +925,8 @@ struct tcpci *tcpci_register_port(struct device *dev, struct tcpci_data *data)
 	tcpci->tcpc.frs_sourcing_vbus = tcpci_frs_sourcing_vbus;
 	tcpci->tcpc.set_partner_usb_comm_capable = tcpci_set_partner_usb_comm_capable;
 
+pr_warn("a1");
+
 	if (tcpci->data->auto_discharge_disconnect) {
 		tcpci->tcpc.enable_auto_vbus_discharge = tcpci_enable_auto_vbus_discharge;
 		tcpci->tcpc.set_auto_vbus_discharge_threshold =
@@ -834,19 +935,27 @@ struct tcpci *tcpci_register_port(struct device *dev, struct tcpci_data *data)
 				   TCPC_POWER_CTRL_BLEED_DISCHARGE);
 	}
 
+pr_warn("a2");
+
 	if (tcpci->data->vbus_vsafe0v)
 		tcpci->tcpc.is_vbus_vsafe0v = tcpci_is_vbus_vsafe0v;
 
+pr_warn("a3");
+
 	err = tcpci_parse_config(tcpci);
 	if (err < 0)
 		return ERR_PTR(err);
 
+pr_warn("a4");
+
 	tcpci->port = tcpm_register_port(tcpci->dev, &tcpci->tcpc);
 	if (IS_ERR(tcpci->port)) {
 		fwnode_handle_put(tcpci->tcpc.fwnode);
 		return ERR_CAST(tcpci->port);
 	}
 
+pr_warn("a5");
+
 	return tcpci;
 }
 EXPORT_SYMBOL_GPL(tcpci_register_port);
@@ -865,6 +974,8 @@ static int tcpci_probe(struct i2c_client *client,
 	int err;
 	u16 val = 0;
 
+dev_warn(&client->dev, "init");
+
 	chip = devm_kzalloc(&client->dev, sizeof(*chip), GFP_KERNEL);
 	if (!chip)
 		return -ENOMEM;
@@ -881,6 +992,8 @@ static int tcpci_probe(struct i2c_client *client,
 	if (err < 0)
 		return err;
 
+dev_warn(&client->dev, "register port");
+
 	chip->tcpci = tcpci_register_port(&client->dev, &chip->data);
 	if (IS_ERR(chip->tcpci))
 		return PTR_ERR(chip->tcpci);
@@ -895,6 +1008,8 @@ static int tcpci_probe(struct i2c_client *client,
 		return err;
 	}
 
+dev_warn(&client->dev, "set wakeup capable");
+
 	device_set_wakeup_capable(chip->tcpci->dev, true);
 
 	return 0;
diff --git a/drivers/usb/typec/tcpm/tcpm.c b/drivers/usb/typec/tcpm/tcpm.c
index 96bb7e72d376..58fee95af0b6 100644
--- a/drivers/usb/typec/tcpm/tcpm.c
+++ b/drivers/usb/typec/tcpm/tcpm.c
@@ -941,16 +941,25 @@ static int tcpm_mux_set(struct tcpm_port *port, int state,
 	tcpm_log(port, "Requesting mux state %d, usb-role %d, orientation %d",
 		 state, usb_role, orientation);
 
+pr_warn("Requesting mux state %d, usb-role %d, orientation %d",
+		 state, usb_role, orientation);
+
+pr_warn("tcpm_mux_set 1");
+
 	ret = typec_set_orientation(port->typec_port, orientation);
 	if (ret)
 		return ret;
 
+pr_warn("tcpm_mux_set 2");
+
 	if (port->role_sw) {
 		ret = usb_role_switch_set_role(port->role_sw, usb_role);
 		if (ret)
 			return ret;
 	}
 
+pr_warn("tcpm_mux_set 3");
+
 	return typec_set_mode(port->typec_port, state);
 }
 
@@ -1028,6 +1037,8 @@ static int tcpm_set_current_limit(struct tcpm_port *port, u32 max_ma, u32 mv)
 
 static int tcpm_set_attached_state(struct tcpm_port *port, bool attached)
 {
+pr_warn("tcpm_set_attached_state");
+
 	return port->tcpc->set_roles(port->tcpc, attached, port->pwr_role,
 				     port->data_role);
 }
@@ -1039,6 +1050,8 @@ static int tcpm_set_roles(struct tcpm_port *port, bool attached,
 	enum usb_role usb_role;
 	int ret;
 
+pr_warn("tcpm_set_attached_roles");
+
 	if (port->polarity == TYPEC_POLARITY_CC1)
 		orientation = TYPEC_ORIENTATION_NORMAL;
 	else
@@ -1089,14 +1102,20 @@ static int tcpm_set_pwr_role(struct tcpm_port *port, enum typec_role role)
 {
 	int ret;
 
+pr_warn("tcpm_set_pwr_role 1");
+
 	ret = port->tcpc->set_roles(port->tcpc, true, role,
 				    port->data_role);
 	if (ret < 0)
 		return ret;
 
+pr_warn("tcpm_set_pwr_role 2");
+
 	port->pwr_role = role;
 	typec_set_pwr_role(port->typec_port, role);
 
+pr_warn("tcpm_set_pwr_role 3");
+
 	return 0;
 }
 
@@ -3561,6 +3580,8 @@ static int tcpm_set_vbus(struct tcpm_port *port, bool enable)
 {
 	int ret;
 
+pr_warn("tcpm_set_vbus enable %B charge %d", enable, port->vbus_charge);
+
 	if (enable && port->vbus_charge)
 		return -EINVAL;
 
@@ -3578,6 +3599,8 @@ static int tcpm_set_charge(struct tcpm_port *port, bool charge)
 {
 	int ret;
 
+pr_warn("tcpm_set_charge %d", charge);
+
 	if (charge && port->vbus_source)
 		return -EINVAL;
 
@@ -3609,6 +3632,8 @@ static int tcpm_init_vbus(struct tcpm_port *port)
 {
 	int ret;
 
+pr_warn("tcpm_init_vbus");
+
 	ret = port->tcpc->set_vbus(port->tcpc, false, false);
 	port->vbus_source = false;
 	port->vbus_charge = false;
@@ -3683,6 +3708,8 @@ static int tcpm_src_attach(struct tcpm_port *port)
 			goto out_disable_pd;
 	}
 
+pr_warn("1a");
+
 	ret = tcpm_set_vbus(port, true);
 	if (ret < 0)
 		goto out_disable_vconn;
@@ -3740,6 +3767,9 @@ static void tcpm_set_partner_usb_comm_capable(struct tcpm_port *port, bool capab
 
 static void tcpm_reset_port(struct tcpm_port *port)
 {
+pr_warn("tcpm_reset_port - disabled");
+return;
+
 	tcpm_enable_auto_vbus_discharge(port, false);
 	port->in_ams = false;
 	port->ams = NONE_AMS;
@@ -3751,28 +3781,41 @@ static void tcpm_reset_port(struct tcpm_port *port)
 	port->pps_data.supported = false;
 	tcpm_set_partner_usb_comm_capable(port, false);
 
+
 	/*
 	 * First Rx ID should be 0; set this to a sentinel of -1 so that
 	 * we can check tcpm_pd_rx_handler() if we had seen it before.
 	 */
 	port->rx_msgid = -1;
 
+pr_warn("tcpm_reset_port 1");
+
 	port->tcpc->set_pd_rx(port->tcpc, false);
+
+pr_warn("tcpm_reset_port 2")
+;
 	tcpm_init_vbus(port);	/* also disables charging */
+pr_warn("tcpm_reset_port 3");
 	tcpm_init_vconn(port);
+pr_warn("tcpm_reset_port 4");
 	tcpm_set_current_limit(port, 0, 0);
+pr_warn("tcpm_reset_port 5");
 	tcpm_set_polarity(port, TYPEC_POLARITY_CC1);
+pr_warn("tcpm_reset_port 6");
 	tcpm_mux_set(port, TYPEC_STATE_SAFE, USB_ROLE_NONE,
 		     TYPEC_ORIENTATION_NONE);
+pr_warn("tcpm_reset_port 7");
 	tcpm_set_attached_state(port, false);
 	port->try_src_count = 0;
 	port->try_snk_count = 0;
 	port->usb_type = POWER_SUPPLY_USB_TYPE_C;
+pr_warn("tcpm_reset_port 8");
 	power_supply_changed(port->psy);
 	port->nr_sink_caps = 0;
 	port->sink_cap_done = false;
 	if (port->tcpc->enable_frs)
 		port->tcpc->enable_frs(port->tcpc, false);
+pr_warn("tcpm_reset_port 9");
 
 	usb_power_delivery_unregister_capabilities(port->partner_sink_caps);
 	port->partner_sink_caps = NULL;
@@ -3780,6 +3823,7 @@ static void tcpm_reset_port(struct tcpm_port *port)
 	port->partner_source_caps = NULL;
 	usb_power_delivery_unregister(port->partner_pd);
 	port->partner_pd = NULL;
+pr_warn("tcpm_reset_port end");
 }
 
 static void tcpm_detach(struct tcpm_port *port)
@@ -4526,6 +4570,7 @@ static void run_state_machine(struct tcpm_port *port)
 		 * drive VBUS to vSafe0V as shown in Figure 7-9.
 		 */
 		tcpm_set_vconn(port, false);
+pr_warn("1b");
 		tcpm_set_vbus(port, false);
 		tcpm_set_roles(port, port->self_powered, TYPEC_SOURCE,
 			       tcpm_data_role_for_source(port));
@@ -4544,6 +4589,8 @@ static void run_state_machine(struct tcpm_port *port)
 		break;
 	case SRC_HARD_RESET_VBUS_ON:
 		tcpm_set_vconn(port, true);
+pr_warn("1c");
+
 		tcpm_set_vbus(port, true);
 		if (port->ams == HARD_RESET)
 			tcpm_ams_finish(port);
@@ -4729,6 +4776,8 @@ static void run_state_machine(struct tcpm_port *port)
 		 * Prevent vbus discharge circuit from turning on during PR_SWAP
 		 * as this is not a disconnect.
 		 */
+pr_warn("1d");
+
 		tcpm_set_vbus(port, false);
 		port->explicit_contract = false;
 		/* allow time for Vbus discharge, must be < tSrcSwapStdby */
@@ -4779,6 +4828,8 @@ static void run_state_machine(struct tcpm_port *port)
 	case PR_SWAP_SNK_SRC_SOURCE_ON:
 		tcpm_enable_auto_vbus_discharge(port, true);
 		tcpm_set_cc(port, tcpm_rp_cc(port));
+pr_warn("1e");
+
 		tcpm_set_vbus(port, true);
 		/*
 		 * allow time VBUS ramp-up, must be < tNewSrc
@@ -5992,7 +6043,9 @@ static void tcpm_init(struct tcpm_port *port)
 
 	port->tcpc->init(port->tcpc);
 
-	tcpm_reset_port(port);
+pr_warn("tcpm_init");
+
+//	tcpm_reset_port(port);
 
 	/*
 	 * XXX
