diff -ur git.org/arch/arm/dts/imx8mm-evk.dts git/arch/arm/dts/imx8mm-evk.dts
--- git.org/arch/arm/dts/imx8mm-evk.dts	2024-08-19 13:52:30.409928118 +0000
+++ git/arch/arm/dts/imx8mm-evk.dts	2024-08-23 15:57:46.769813439 +0000
@@ -9,7 +9,7 @@
 #include "imx8mm-evk.dtsi"
 
 / {
-	model = "NXP i.MX8MM EVK board";
+	model = "i.MX8MM Jaguar Sentai board";
 	compatible = "fsl,imx8mm-evk", "fsl,imx8mm";
 
 	chosen {
@@ -109,4 +109,19 @@
 			MX8MM_IOMUXC_NAND_CE1_B_USDHC3_STROBE           0x196
 		>;
 	};
+
+        pinctrl_i2c4: i2c4grp {
+                fsl,pins = <
+                        MX8MM_IOMUXC_I2C4_SCL_I2C4_SCL                  0x400001c3
+                        MX8MM_IOMUXC_I2C4_SDA_I2C4_SDA                  0x400001c3
+                >;
+        };
+};
+
+&i2c4 {
+        clock-frequency = <400000>;
+        pinctrl-names = "default";
+        pinctrl-0 = <&pinctrl_i2c4>;
+
+        status = "okay";
 };
