diff --git a/arch/arm64/boot/dts/freescale/imx8mm-evkb.dts b/arch/arm64/boot/dts/freescale/imx8mm-evkb.dts
index 86922c452066..0b46e89579ef 100755
--- a/arch/arm64/boot/dts/freescale/imx8mm-evkb.dts
+++ b/arch/arm64/boot/dts/freescale/imx8mm-evkb.dts
@@ -161,3 +161,45 @@ MX8MM_IOMUXC_SD1_DATA7_GPIO2_IO9		0x159
 		>;
 	};
 };
+
+&adv_bridge {
+        status = "disabled";
+};
+
+&mipi_dsi {
+        panel@0 {
+                compatible = "sitronix,st7701";
+                reg = <0>;
+                pinctrl-0 = <&pinctrl_mipi_dsi_en>;
+                reset-gpio = <&gpio1 8 GPIO_ACTIVE_HIGH>;
+                dsi-lanes = <2>;
+                video-mode = <2>;       /* 0: burst mode
+                                         * 1: non-burst mode with sync event
+                                         * 2: non-burst mode with sync pulse
+                                         * 3: command mode
+                                         */
+                panel-width-mm = <36>;
+                panel-height-mm = <65>;
+                status = "okay";
+        };
+};
+
+&i2c2 { 
+        clock_frequency = <100000>;
+        pinctrl-names = "default";
+        pinctrl-0 = <&pinctrl_i2c2>;
+        status = "okay";
+
+        polytouch: edt-ft5x06@38 {
+                compatible = "edt,edt-ft5406", "edt,edt-ft5x06";
+                reg = <0x38>;
+                pinctrl-names = "default";
+                pinctrl-0 = <&pinctrl_i2c2_synaptics_dsx_io>;
+                interrupt-parent = <&gpio1>;
+                interrupts = <9 IRQ_TYPE_LEVEL_LOW>;
+
+                touchscreen-size-x = <360>;
+                touchscreen-size-y = <640>;
+        };
+};
+
